docs/proposals/24-dynamic-node-groups-for-shoot-cluster-autoscaler.md

```diff
@@ -0,0 +1,338 @@
+    - [2. Node Group Auto Discovery](#2-node-group-auto-discovery)
+    - [3. Dynamic Node Group Sizing.](#3-dynamic-node-group-sizing)
+    - [3.1 Lax-Greedy Sizing](#31-lax-greedy-sizing)
+      - [Pros/Cons](#proscons)
+      - [Examples](#examples)
+        - [Pool-3:4:2:2:3 (min:max:maxsurge:maxunavail:numzones)](#pool-34223-minmaxmaxsurgemaxunavailnumzones)
+          - [Scan-0](#scan-0)
+          - [Scan-1](#scan-1)
+          - [Scan-2](#scan-2)
+          - [Scan-3](#scan-3)
+    - [3.2 Backward-Compatible](#32-backward-compatible)
+      - [Pros/Cons](#proscons-1)
+      - [Examples](#examples-1)
+        - [Pool-3:4:2:2:3 (min:max:maxsurge:maxunavail:numzones)](#pool-34223-minmaxmaxsurgemaxunavailnumzones-1)
+          - [Scan-0](#scan-0-1)
+          - [Scan-1](#scan-1-1)
+          - [Scan-2](#scan-2-1)
+          - [Pool-0:2:1:1:2 (min:max:maxsurge:maxunavail:numzones)](#pool-02112-minmaxmaxsurgemaxunavailnumzones)
+          - [Scan-0](#scan-0-2)
+          - [Scan-1](#scan-1-2)
+    - [4. Shoot Spec Enhancement](#4-shoot-spec-enhancement)
+    - [TODO](#todo)
+
+## Summary
+
+Presently, the `NodeGroups` for the shoot `cluster-autoscaler`  are given static `Minimum` and `Maximum` sizes computed in the gardener extension providers during shoot reconciliation. This has several disadvantages since the static computation preferences zones that are specified earlier over zones specified later. In addition, for cases where the number of zones exceeds the `Max` specified for the worker pool, the node-group is assigned a zero `Max` size.
+
+This document outlines deficiences in the existing solution and proposes moving to the CA's node group auto-discovery mechanism where node groups can be discovered and sized at runtime. 
+
+## Motivation
+
+TODO: Add Sequence Diagram for the below.
+
+Let us first study how `MachineDeployments` are currently generated and how node groups are currently computed for the `cluster-autoscaler`.
+
+
+### Existing Solution
+
+#### Worker Actuator Reconciliation 
+
+The Worker [generic.Actuator.Reconcile](https://github.com/gardener/gardener/blob/a3632ea5315d104d0ed3dd47e6d17f53cfe0e877/extensions/pkg/controller/worker/genericactuator/actuator_reconcile.go#L42) implementation delegates to [genericactuator.GenerateMachineDeployments](https://github.com/gardener/gardener/blob/a3632ea5315d104d0ed3dd47e6d17f53cfe0e877/extensions/pkg/controller/worker/genericactuator/interface.go#L46) to generate [worker.MachineDeployments](https://github.com/gardener/gardener/blob/a3632ea5315d104d0ed3dd47e6d17f53cfe0e877/extensions/pkg/controller/worker/machines.go#L39) for a shoot clulster during shoot reconciliation.
+
+These Worker MachineDeployments's are converted into MCM [MachineDeployments](https://pkg.go.dev/github.com/gardener/machine-controller-manager@v0.50.0/pkg/apis/machine/v1alpha1#MachineDeployment)'s and then deployed into the shoot control plane.
```

Comment:
Maybe you can also include the link where this is done.
https://github.com/gardener/gardener/blob/master/extensions/pkg/controller/worker/genericactuator/actuator_reconcile.go#L204 is the method where this is done. This only helps the reviewer to quickly navigate to the code where this is currently done.